### PR TITLE
bots: Adjust atomic.setup for latest Fedora 27 Atomic

### LINKED ISSUE #8691: Image refresh for fedora-atomic
Image refresh for fedora-atomic
 * [ ] FAIL: image-refresh fedora-atomic

### PR DESCRIPTION
**Pull request by martinpitt**
Latest Fedora Atomic changes the docker storage LVM configuration; not in a good way IMHO, but we have to deal with it. See individual commit logs.

Fixes #8691

 * [x] image-refresh fedora-atomic

### CODE REVIEW

**Comment by cockpituous** [comment]
image-refresh in progress on cockpit-tasks-639m0.
Log: http://fedorapeople.org/groups/cockpit/logs/image-refresh-8694-20180223-082727/

**Comment by cockpituous** [comment]
image-refresh fedora-atomic done: https://github.com/cockpituous/cockpit/commits/image-refresh-fedora-atomic-20180223-084040

**Comment by martinpitt** [comment]
I pushed the built image.

**Comment by martinpitt** [comment]
There are two [relevant test failures](https://fedorapeople.org/groups/cockpit/logs/pull-8694-20180223-084343-2f77673c-verify-fedora-atomic/log.html). I fixed check-metrics locally, the docker-storage pool one gives me some headaches.

**Comment by martinpitt** [comment]
So the issue is that latest docker images don't create a separate docker LV any more:
```
# cat /etc/sysconfig/docker-storage-setup 
# Edit this file to override any configuration options specified in
# /usr/share/container-storage-setup/container-storage-setup.
#
# For more details refer to "man container-storage-setup"
STORAGE_DRIVER=overlay2
GROWPART=true
ROOT_SIZE=+100%FREE
```

Previous releases had this:
```
# cat /etc/sysconfig/docker-storage-setup 
# Edit this file to override any configuration options specified in
# /usr/share/container-storage-setup/container-storage-setup.
#
# For more details refer to "man container-storage-setup"
STORAGE_DRIVER=overlay2
GROWPART=true
CONTAINER_ROOT_LV_NAME=docker-root-lv
CONTAINER_ROOT_LV_MOUNT_PATH=/var/lib/docker
```

I've tried to force this back to the previous settings:
```patch
--- a/bots/images/scripts/fedora-atomic.setup
+++ b/bots/images/scripts/fedora-atomic.setup
@@ -8,6 +8,18 @@ set -ex
 printf '(allow init_t domain (process (rlimitinh)))\n' > domain.cil
 semodule -i domain.cil
 
+# use separate LV for docker
+if ! grep -q docker-root-lv /etc/sysconfig/docker-storage-setup; then
+    sed -i '/^ROOT_SIZE/d' /etc/sysconfig/docker-storage-setup
+
+    cat <<EOF >> /etc/sysconfig/docker-storage-setup
+CONTAINER_ROOT_LV_NAME=docker-root-lv
+CONTAINER_ROOT_LV_MOUNT_PATH=/var/lib/docker
+EOF
+    lvresize --force -l50%VG /dev/atomicos/root
+    docker-storage-setup
+fi
+
 # HACK: docker falls over regularly, print its log if it does
 systemctl start docker || journalctl -u docker
```
But this still fails, as this will override any existing images in /var/lib/docker (`open /var/lib/docker/tmp/GetImageBlob492861700: no such file or directory`); plus, the forced shrinking of the root partition is terminally unsafe as it cannot shrink the file system in it (XFS does not suppport that).

So I think we have to live with this Fedora Atomic decision and just not support doing docker-storage-setup there. Unfortunately the test still sees the cd-rom drive and considers that a storage device (which is a bug in itself), so it doesn't show the "unsupported" message (`#storage-unsupported`) and thus we can't just add it to `can_manage()` - we also need to fix that bug to not show the cloud-init ISO CD.

**Comment by martinpitt** [comment]
When I step through manually, I see that resetting the storage pool in the test does not actually reset it - after that, the displayed storage pool still includes DISK1 and DISK2, they are not in the "additional storage" where the test (rightfully) expects them. When I click the button manually, this is still the same result, so reasonably well reproducible.
 
That actually might be related to the Fedora change - DISK1 and DISK2 get added as PVs to the root atomicos VG, so they might not actually be separable from docker any more.

**Comment by martinpitt** [comment]
I think I fully understand what's going on now, and I adjusted the test. I'm not completely happy about this, but maybe @mvollmer has a better idea.

**Comment by martinpitt** [comment]
Tests fail on centos/rhel 7, apparently they don't have initial docker images? Investigating..

**Comment by mvollmer** [comment]
It's been a while, but I think we (try to) suport this scenario.  Adding a disk should create the volume group.  I'll check this.

**Comment by martinpitt** [comment]
> Adding a disk should create the volume group. I'll check this.

Except that it can't, as there is no more available space, and shrinking XFS isn't supported. See comment above.

**Comment by mvollmer** [comment]
>> Adding a disk should create the volume group. I'll check this.

> Except that it can't, as there is no more available space, and shrinking XFS isn't supported. See comment above.

Well, there is space on the added disk.  So, "adding" might actually make the pool smaller...

But in any case, I misunderstood the scenario; I thought there isn't any volume group to start with.

I check what is going wrong with "reset".

**Comment by martinpitt** [comment]
Gah @ https://fedorapeople.org/groups/cockpit/logs/pull-8694-20180226-145151-c8b1248f-verify-fedora-atomic/log.html#148 : this needs another naughty override. It's a flake, but let's fix it while I'm at it.

File: test/verify/check-docker-storage [hunk 1]
```diff
@@ -228,6 +231,10 @@ class TestDockerStorageMapper(MachineCase):
         b.wait_not_present(".modal-dialog:contains(Add Additional Storage)")
         b.wait_not_in_text("#storage-drives", "DISK2")
 
+        # adding disks should keep images, unless the above created the VG and thus reset the pool
+        if not initially_without_vgroup(m):
```
Comment (mvollmer) [inline test/verify/check-docker-storage]: The full condition is `initially_loopbacked(m) or initially_without_vgroup(m)`.

**Comment by mvollmer** [comment]
I added a commit that removes the "reset" button completely when it doesn't work as expected.

**Comment by martinpitt** [comment]
> I added a commit that removes the "reset" button completely when it doesn't work as expected.

So this is meant to completely remove the VG, not just clean up all docker containers and images? Because it does the latter just fine. But I'm OK with that approach.

> The full condition is initially_loopbacked(m) or initially_without_vgroup(m).

Done and pushed.

**Comment by mvollmer** [comment]
> So this is meant to completely remove the VG, not just clean up all docker containers and images? 

It is supposed to give you back the disks that you had previously added.  Removing containers and images is our big hammer to make that happen.

I start to think that we should just give people the normal storage page instead of this docker specific subset.

**Comment by martinpitt** [comment]
Test now fails on rhel-7-5, which doesn't have a Reset button, but the test expects one:
https://fedorapeople.org/groups/cockpit/logs/pull-8694-20180227-140606-5a60fa60-verify-rhel-7-5/log.html#138

The other failures look flake-ish to me.

**Comment by martinpitt** [comment]
On RHEL 7.5, the docker storage is also on the root LV:
```
# mount|grep docker
/dev/mapper/rhel-root on /var/lib/docker/containers type xfs (rw,relatime,seclabel,attr2,inode64,noquota)
/dev/mapper/rhel-root on /var/lib/docker/overlay2 type xfs (rw,relatime,seclabel,attr2,inode64,noquota)

# lvs
  LV   VG   Attr       LSize   Pool Origin Data%  Meta%  Move Log Cpy%Sync Convert
  root rhel -wi-ao----  <6.20g                                                    
  swap rhel -wi-ao---- 820.00m                   
```
so our monitor rightfully claims that it cannot reset:
```
# /tmp/cockpit-atomic-storage monitor
{"vgroup": "rhel", "can_manage": true, "total": 2449567744, "pool_devices": [{"name": "QEMU QEMU_HARDDISK (DISK1)", "sort_index": 800000, "path": "/dev/sda1", "shared": false, "class": "hdd", "size": 104857600}, {"name": "QEMU QEMU_HARDDISK (DISK2)", "sort_index": 800016, "path": "/dev/sdb1", "shared": false, "class": "hdd", "size": 104857600}, {"name": "Virtio Disk (ROOT)", "sort_index": 25200000, "path": "/dev/vda2", "shared": true, "class": "hdd", "size": 8589934592}], "loopback": false, "can_reset": false, "used": 0, "extra_devices": [{"path": "/dev/sr0", "class": "drive", "size": 374784, "name": "QEMU_DVD-ROM (QM00002)", "sort_index": 1100000}]}
```

I think this is correct, and therefore we should add rhel-7-5 to the `can_reset()` list in the test. Fix pushed.

**Pull request merged**